### PR TITLE
RFC: Process sceSas in blocks of 32 samples

### PR DESCRIPTION
**Pull request by hrydgard**
As we talked about, reduces the amount of CPU work in audio ADSR curve walking. Probably a very slight speedup.

However, the adsrcurve test is very slightly off... so probably shouldn't merge this as is.

### CODE REVIEW

**Comment by daniel229** [comment]
Breaks Senjou no Valkyria 3,no sound.

File: Core/HW/SasAudio.cpp [hunk 1]
```diff
@@ -782,27 +787,38 @@ void ADSREnvelope::SetState(ADSRState state) {
...
 }
 
+void ADSREnvelope::GetTypeAndRate(int &type, int &rate) {
+	switch (state_) {
+	case STATE_ATTACK: type = attackType; rate = attackRate; break;
+	case STATE_DECAY: type = decayType; rate = decayRate; break;
+	case STATE_SUSTAIN: type = sustainType; rate = sustainRate; break;
+	case STATE_RELEASE: type = releaseType; rate = releaseRate; break;
+	case STATE_KEYON:
+		height_ = 0;
```
Comment (unknownbrackets) [inline Core/HW/SasAudio.cpp]: Seems like this should return some kind of type/rate (possibly -1 or something?)

-[Unknown]

Comment (unknownbrackets) [inline Core/HW/SasAudio.cpp]: Or, actually, maybe this:
```c++

			if (!voice.envelope.GetTypeAndRate(type, rate)) {
				voice.envelope.Step();
				continue;
			}
```

-[Unknown]

**Comment by unknownbrackets** [comment]
Hmm, yeah, so in Senjou no Valkyria 3, `voice.envelope.Step(i - j);` shows that it switches over on step 4/5 (`i - j` = 3 and 4.)  Hmm.  I suppose it doesn't do it in block after all, sorry... my bad.

-[Unknown]

**Pull request closed without merging**